src/mjolnir/Mjolnir.js

```diff
@@ -0,0 +1,189 @@
+                topic: _t("This is your list of users/servers you have blocked - don't leave the room!"),
+                preset: "private_chat",
+            });
+            personalRoomId = resp['room_id'];
+            await SettingsStore.setValue(
+                "mjolnirPersonalRoom", null, SettingLevel.ACCOUNT, personalRoomId);
+            await SettingsStore.setValue(
+                "mjolnirRooms", null, SettingLevel.ACCOUNT, [personalRoomId, ...this._roomIds]);
+        }
+        if (!personalRoomId) {
+            throw new Error("Error finding a room ID to use");
+        }
+
+        let list = this._lists.find(b => b.roomId === personalRoomId);
+        if (!list) list = new BanList(personalRoomId);
+        // we don't append the list to the tracked rooms because it should already be there.
+        // we're just trying to get the caller some utility access to the list
+
+        return list;
+    }
+
+    // get without creating the list
+    getPersonalList(): BanList {
+        const personalRoomId = SettingsStore.getValue("mjolnirPersonalRoom");
+        if (!personalRoomId) return null;
+
+        let list = this._lists.find(b => b.roomId === personalRoomId);
+        if (!list) list = new BanList(personalRoomId);
+        // we don't append the list to the tracked rooms because it should already be there.
+        // we're just trying to get the caller some utility access to the list
+
+        return list;
+    }
+
+    async subscribeToList(roomId: string) {
+        const roomIds = [...this._roomIds, roomId];
+        await SettingsStore.setValue("mjolnirRooms", null, SettingLevel.ACCOUNT, roomIds);
+        this._lists.push(new BanList(roomId));
+    }
+
+    async unsubscribeFromList(roomId: string) {
+        const roomIds = this._roomIds.filter(r => r !== roomId);
+        await SettingsStore.setValue("mjolnirRooms", null, SettingLevel.ACCOUNT, roomIds);
+        this._lists = this._lists.filter(b => b.roomId !== roomId);
+    }
+
+    _onEvent(event) {
+        if (!MatrixClientPeg.get()) return;
+        if (!this._roomIds.includes(event.getRoomId())) return;
+        if (!ALL_RULE_TYPES.includes(event.getType())) return;
+
+        this._updateLists(this._roomIds);
+    }
+
+    _onListsChanged(settingName, roomId, atLevel, newValue) {
+        // We know that ban lists are only recorded at one level so we don't need to re-eval them
+        this._updateLists(newValue);
+    }
+
+    _updateLists(listRoomIds: string[]) {
+        if (!MatrixClientPeg.get()) return;
+
+        console.log("Updating Mjolnir ban lists to: " + listRoomIds);
+        this._lists = [];
+        this._roomIds = listRoomIds || [];
+        if (!listRoomIds) return;
+
+        for (const roomId of listRoomIds) {
+            // Creating the list updates it
+            this._lists.push(new BanList(roomId));
+        }
+    }
+
+    isServerBanned(serverName: string): boolean {
+        for (const list of this._lists) {
```

Comment:
wondering whether this and `isUserBanned` are going slow things down while rendering the timeline (we re-render the timeline often I think).

Reply:
it's highly cached data, so shouldn't be too bad. The globs also make the regex small and fast, so shouldn't be too bad. Running it on my account with a lengthy list had no noticeable impact.

Reply:
Hmm, I guess fine for now as it's behind a labs flag. Would be good to get some numbers if we enable this by default though.

Reply:
do we even have a system for counting time-based metrics like this?

Reply:
also I'd probably want to do it as its own PR given the size of this one.